### PR TITLE
chore: use the Engine API from containerd-shim-wasm crate

### PR DESCRIPTION
**Pull request by Mossaka**
This commits updates to use the latest API from the `containerd-shim-wasm` crate. It follows the PR https://github.com/containerd/runwasi/pull/293

FYI: @jprendes

### CODE REVIEW

File: containerd-shim-wws-v1/src/executor.rs [hunk 1]
```diff
@@ -48,39 +61,11 @@ impl WwsExecutor {
...
-            if let Err(err) = self.wasm_exec(spec) {
-                log::info!(" >>> server shut down due to error: {err}");
-                std::process::exit(137);
-            }
-            log::info!(" >>> server shut down: exiting");
-            std::process::exit(0);
+        if let Err(e) = rt.block_on(self.wasm_exec_async(path, routes)) {
+            log::error!(" >>> error: {:?}", e);
+            return Ok(137);
         }
```
Comment (jprendes) [inline containerd-shim-wws-v1/src/executor.rs:68]:
```suggestion
        rt.block_on(self.wasm_exec_async(path, routes))?
```

**Review by jprendes — CHANGES_REQUESTED** [review]
Most of the comments are just suggestion. The main thing that *needs* fixing is that the default `can_handle` is not suitable for most of the shims (it might be ok for lunatic).

Could you rename the `executor.rs` files to `engine.rs`?

I think many of the crates dependencies can be dropped now, just to take some examples from lunatic's Cargo.toml:
* libcontainer
* nix
* serde
* serde_json
* thiserror
* libc
* chrono
* oci-spec
* utils
* clap (needs removing the `derive(Args)` in `common.rs`)

But all shims could use a little clean up.

Also I believe the `utils` crate be removed.

File: containerd-shim-lunatic-v1/src/executor.rs [hunk 2]
```diff
@@ -3,58 +3,38 @@ use std::path::PathBuf;
...
-        let cmd = get_args(spec).first().context("no cmd provided")?.clone();
+    fn run_wasi(&self, ctx: &impl RuntimeContext, stdio: Stdio) -> Result<i32> {
+        log::info!("setting up wasi");
+        stdio.redirect()?;
+        let cmd = ctx.args().first().context("no cmd provided")?.clone();
         let rt = Runtime::new().context("failed to create runtime")?;
-        rt.block_on(async {
+        if let Err(e) = rt.block_on(async {
             log::info!(" >>> building lunatic application");
             crate::executor::exec(cmd).await
```
[the remark this answers is not in the document]
Comment (Mossaka) [reply containerd-shim-lunatic-v1/src/executor.rs:32]: Done.

File: containerd-shim-slight-v1/src/executor.rs [hunk 3]
```diff
@@ -1,29 +1,23 @@
 use anyhow::{Context, Result};
+use containerd_shim_wasm::container::RuntimeContext;
 use std::path::PathBuf;
 use tokio::runtime::Runtime;
 
-use containerd_shim_wasm::libcontainer_instance::LinuxContainerExecutor;
+use containerd_shim_wasm::container::Engine;
 use containerd_shim_wasm::sandbox::Stdio;
```
[the remark this answers is not in the document]
Comment (Mossaka) [reply containerd-shim-slight-v1/src/executor.rs:7]: Will follow up with another PR

File: containerd-shim-lunatic-v1/src/executor.rs [hunk 4]
```diff
@@ -3,58 +3,38 @@ use std::path::PathBuf;
...
-pub struct LunaticExecutor {
-    stdio: Stdio,
-}
+#[derive(Clone, Default)]
+pub struct LunaticEngine;
 
-impl LunaticExecutor {
-    pub fn new(stdio: Stdio) -> Self {
-        Self { stdio }
+impl Engine for LunaticEngine {
```
[the remark this answers is not in the document]
Comment (Mossaka) [reply containerd-shim-lunatic-v1/src/executor.rs:17]: Yes, lunatic shim requires a wasm entrypoint

File: containerd-shim-spin-v1/src/executor.rs [hunk 5]
```diff
@@ -111,22 +95,44 @@ impl SpinExecutor {
...
-            log::info!("executing linux container");
-            LinuxContainerExecutor::new(self.stdio.clone()).exec(spec)
-        } else {
-            if let Err(err) = self.wasm_exec(spec) {
-                log::info!(" >>> server shut down due to error: {err}");
-                std::process::exit(137);
-            }
-            log::info!(" >>> server shut down: exiting");
-            std::process::exit(0);
+impl Engine for SpinEngine {
```
[the remark this answers is not in the document]
Comment (Mossaka) [reply containerd-shim-spin-v1/src/executor.rs:98]: Done

File: containerd-shim-wws-v1/src/executor.rs [hunk 6]
```diff
@@ -3,10 +3,10 @@ use log::{error, info};
...
 use tokio::runtime::Runtime;
 
-use containerd_shim_wasm::{libcontainer_instance::LinuxContainerExecutor, sandbox::Stdio};
-use libcontainer::workload::{Executor, ExecutorError, ExecutorValidationError};
-use oci_spec::runtime::Spec;
-use utils::is_linux_executable;
+use containerd_shim_wasm::{
+    container::{Engine, RuntimeContext},
+    sandbox::Stdio,
+};
```
[the remark this answers is not in the document]
Comment (Mossaka) [reply containerd-shim-wws-v1/src/executor.rs:9]: Done.

**Comment by Mossaka** [comment]
Thanks for your careful review of this PR! @jprendes 

I have resolved all your comments.

**Review by jprendes — APPROVED** [review]
LGTM!

**Pull request merged**